segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ReadOperation.java

```diff
@@ -141,23 +143,56 @@ private void logEnd() {
                     } else {
                         Preconditions.checkState(bytesToRead != 0, "bytesToRead is 0");
                         // Read data from the chunk.
-                        return chunkedSegmentStorage.getChunkStorage().openRead(chunkToReadFrom.getName())
-                                .thenComposeAsync(chunkHandle ->
-                                        chunkedSegmentStorage.getChunkStorage().read(chunkHandle,
-                                                currentOffset.get() - startOffsetForCurrentChunk.get(),
-                                                bytesToRead,
-                                                buffer,
-                                                currentBufferOffset.get())
-                                                .thenAcceptAsync(bytesRead -> {
-                                                    bytesRemaining.addAndGet(-bytesRead);
-                                                    currentOffset.addAndGet(bytesRead);
-                                                    currentBufferOffset.addAndGet(bytesRead);
-                                                    totalBytesRead.addAndGet(bytesRead);
-                                                }, chunkedSegmentStorage.getExecutor()),
-                                        chunkedSegmentStorage.getExecutor());
+                        return CompletableFuture.runAsync(() -> {
+                            // Create parallel requests to read each chunk.
+                            chunkReadFutures.add(readChunk(chunkToReadFrom.getName(),
```

Comment:
Issuing read requests serially is bad for performance, but it seems harder to overload the cache upon the arrival of many read requests of large chunks. That is, what happens if the size of the cache is smaller than a set of read request that are sent in parallel? Could it be the case that the storage reads keep getting `CacheFullException` because this is just reading in parallel as fast as possible from storage without control? (in a similar way in which we got `CacheFullExceptions` when recovering containers in parallel without controlling the number of parallel recoveries). What about having a number of parallel requests > the storage pool size? Is this going to starve data movements from `StorageWriter` to tier 2?

Reply:
It is the cache that issued the original read request. We are just using parallel reads to fulfill that same original resuest. This change should affect nothing as far as cache is concerned.